src/iterators.jl

```diff
@@ -0,0 +1,301 @@
+                    v[i] != (v[i] + 1) && _update_indexs!(inds,v[i], v[i] + 1) 
+                    v[i] = (v[i] + 1)
+                else
+                    v[i] != i && _update_indexs!(inds, v[i], i) 
+                    v[i] = i
+                end
+            else
+                if ( v[i + 1] != (v[i] + 1) )
+                    if i > 1
+                        v[i - 1] != v[i] && _update_indexs!(inds, v[i - 1], v[i]) 
+                        v[i - 1] = v[i]
+                    end
+                    v[i] != (v[i] + 1) && _update_indexs!(inds, v[i], v[i] + 1) 
+                    v[i] = (v[i] + 1)
+                else
+                    v[i + 1] != v[i] && _update_indexs!(inds, v[i + 1], v[i]) 
+                    v[i + 1] = v[i]
+                    v[i] != i && _update_indexs!(inds, v[i], i) 
+                    v[i] = i
+                end
+            end
+        end
+    end
+    return (inds, (v, rank, inds))
+end
+
+@inline function rest(G::SubsetGrayCode, rank)
+  #TODO there will be a future PR for integrating this (its used for multithreaded weight calculation)
+  #=
+  _subset_unrank(rank, G.n, vec)
+  inds = Vector{Int}([-1,-1,-1]) 
+  state = (vec, rank, inds)
+  return Base.rest(G, state)
+  =#
+end 
+
+function _update_indexs!(indexs::Vector{Int}, x::Int, y::Int)
+    _update_indexs!(indexs, x)
+    _update_indexs!(indexs, y)
+    return nothing
+end
+
+function _update_indexs!(indexs::Vector{Int}, x::Int)
+    if x == indexs[1]
+        indexs[1] = -1
+    elseif x == indexs[2]
+        indexs[2] = -1
+    elseif x == indexs[3]
+        indexs[3] = -1
+    elseif indexs[1] == -1
+        indexs[1] = x
+    elseif indexs[2] == -1
+        indexs[2] = x
+    elseif indexs[3] == -1
+        indexs[3] = x
+    else 
+        throw("No index positions remaining")
+    end
+    return nothing
+end
+
+function _subset_rank(v::Vector{UInt}, k::UInt)
+    #=
+    Based on Algorithm 2.11 in kreher1999combinatorial
+    Results are undefined if the entries of v arent in {1,..,n} for n>=k
+    =#
+    r = BigInt(0)
+    s = BigInt(1)
+    for i in k:-1:1
+        r = r + extended_binomial(v[i], i) * s;
+        s = -s;
+    end
+    if (k % 2) == 1
+        r = r - 1
+    end
+    return r
+end
+
+function _subset_unrank(r::BigInt, n::UInt, T::Vector{UInt})
+    # Based on Algorithm 2.12 in kreher1999combinatorial
+    k = length(T)
+    subset_size_str="subset size k=($k) must be smaller than the set size n=($n)"
+    k > n && throw(ArgumentError(subset_size_str))
+    bnd = binomial(n, k)
+    rank_size_str="rank must be in [0, choose(n, k)-1]=$(bnd)"
+    r > bnd && throw(ArgumentError(rank_size_str))
+  
+    x = 0
+    i = 0
+    y = 0
+    x = n
+    for i::UInt in k:-1:1
+        y = extended_binomial(x, i)
+        while y > r
+           x = x - 1
+           y = extended_binomial(x, i)
+        end 
+        T[i] = x + 1
+        r = extended_binomial(x + 1, i) - r - 1
+    end 
+end
+
+struct GrayCode
```

Comment:
I moved GrayCode. I did not change anything except for adding one reference to the paper the algorithm was initially based on